idxmgmt/ilm/config.go

```diff
@@ -18,34 +18,117 @@
 package ilm
 
 import (
-	"github.com/elastic/beats/libbeat/common/fmtstr"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
 	libilm "github.com/elastic/beats/libbeat/idxmgmt/ilm"
 )
 
-const pattern = "000001"
-
-// Config for ILM supporter
+//Config holds information about ILM mode, overwriting and policies
 type Config struct {
-	Mode       libilm.Mode               `config:"enabled"`
-	PolicyName *fmtstr.EventFormatString `config:"policy_name"`
-	AliasName  *fmtstr.EventFormatString `config:"alias_name"`
-	Event      string                    `config:"event"`
+	Mode          libilm.Mode `config:"enabled"`
+	Overwrite     bool        `config:"overwrite"`
+	RequirePolicy bool        `config:"require_policy"`
+	Policies      []Policy
 }
 
-// Enabled indicates whether or not ILM should be enabled
+//Policy binds together an ILM policy's name and body with an event type
+type Policy struct {
+	Policy    map[string]interface{}
+	EventType string
+	Name      string
+}
+
+//Enabled indicates whether or not ILM should be enabled
 func (c *Config) Enabled() bool {
 	return c.Mode != libilm.ModeDisabled
 }
 
-// ModeString stringifies enabled mode
-// This is a workaround as strings should be generated in libbeat.
-func ModeString(m libilm.Mode) string {
-	switch m {
-	case libilm.ModeAuto:
-		return "auto"
-	case libilm.ModeEnabled:
-		return "true"
-	default:
-		return "false"
+//NewConfig returns an ILM config, where default configuration is merged with user configuration
+func NewConfig(ucfg *common.Config) (Config, error) {
+	policies := policyPool()
+	policyMappings := policyMapping()
+	c := defaultConfig()
+	if ucfg != nil {
+		if err := ucfg.Unpack(&c); err != nil {
+			return Config{}, err
+		}
+
+		var cfg config
+		if err := ucfg.Unpack(&cfg); err != nil {
+			return Config{}, err
+		}
+		// create a collection of default and configured policies
+		for name, policy := range cfg.Policies {
+			policies[name] = policy
+		}
+		//update policy name per event according to configuration
+		for _, entry := range cfg.Mapping {
+			if _, ok := policyMappings[entry.Event]; !ok {
+				return c, errors.Errorf("event_type '%s' not supported for ILM setup", entry.Event)
+			}
+			policyMappings[entry.Event] = entry.Policy
+		}
+	}
+
+	for event, policyName := range policyMappings {
```

Comment:
as counterpoint to https://github.com/elastic/apm-server/pull/2764/files#diff-b9f51268216554aa54cd5bdb34d4ecb1R210, this reads well, much better that if you were to do `for k, v :=...`